### ISSUE TITLE
Getting Target closed error while switching to iframe

### ISSUE THREAD

**Issue by ddeogharkar** [microsoft/playwright#10911]
Hi, 
I am switching to iFrame using latest playwright version 1.17 but I am getting Target closed error. I am using below code 

await this.page.frameLocator("//iframe[@class='hero-card']").locator("[aria-label='Go to Page 4']").click();

Error is :
frame.click: Target closed

    =========================== logs ===========================
    waiting for selector "[aria-label='Go to Page 4']"
    ============================================================
I tried multiple things but it is not working. I am using POM.

**Comment by mxschmitt** [comment]
Could you provide us a full repro? Otherwise it's hard for us to act on it. For now it can be caused by multiple reasons, e.g. the test timeout is reached which closes the page and ends up canceling the `frame.click` call.

**Comment by ddeogharkar** [comment]
Hi,
I am able to resolve this problem, but now I am facing another issue i.e.

When I am using test.beforeAll like below, browser opens but it gets closed immediately without launching url & further process.

import test, { Page } from "@playwright/test";
import { LoginPage } from "../Pages/pageRepository/LoginPage";
import testData from "../Data/appData.json"

let page: Page;
let loginPage: LoginPage;

test.beforeAll(async ({ browser }) => {

  page = await browser.newPage();

  loginPage = new LoginPage(page);

  await loginPage.NavigateToUrl(testData.homePageUrl);

  await loginPage.loginToApplication(testData.stageUserName, testData.password, testData.prodUserName);

  console.log("Execute before all tests");
})

=========================
locator.click: Target closed
    =========================== logs ===========================
      waiting for frame "xpath=//iframe[@class='hero-card']"
    ============================================================

      25 |
      26 |     test.step("Test_04_TopNews_Verify_After_Clicking_Article_NavigateTo_Article_LandingPage", async () => {
    > 27 |       await topNewsPage.clickTopNewsDots();
         |                         ^
      28 |       console.log("Test 2");
      29 |     })
      30 |

**Comment by mxschmitt** [comment]
You don't provide us the full code where the error happens, but it looks like you are missing the `await` before `test.step`.

**Comment by ddeogharkar** [comment]
Please find the git repo for the full code. I have added await before test.step but still I am facing issue with before all.

https://github.com/ddeogharkar/mc_changes

**Comment by Evilweed** [comment]
@mxschmitt we have the same problem. This is internal app that we cannot easily provide access to, and saving it to hard drive does not help - it does not render properly either way (missing iframes/JS that detects tampering and does not show anything).

But we are happy to provide slightly redacted logs.

**Logs showing the issue:**
[log.txt](https://github.com/microsoft/playwright/files/7746187/log.txt)

**Comment by Evilweed** [comment]
Looking at the callstack i think this is another issue related to Websockets.

![CleanShot 2021-12-20 at 16 30 59@2x](https://user-images.githubusercontent.com/5477821/146792323-6f4e73a9-ab89-41aa-81eb-ae5b129cd3aa.png)

![CleanShot 2021-12-20 at 16 32 01@2x](https://user-images.githubusercontent.com/5477821/146792552-888ce3b6-04f4-43b8-a7c0-220ffa9867bb.png)

**Comment by Evilweed** [comment]
Last time we had issues with Websocket+Iframe it was this one https://github.com/microsoft/playwright/pull/10518

**Comment by Evilweed** [comment]
So previous log was when we were clicking on input button, that triggers downloading of missing JS to show date picking modal. We have used page.eval to set the input value without clicking input. This worked but we are still blocked as trying to send the form after clicking submit also crashes Playwright - with following log output:
```
  ✘  admin/custom-offer/add/adding-custom-offer.spec.ts:31:5 › New Custom Offer › should be saved with proper data (16s)
    pw:channel:event {
  pw:channel:event   guid: 'frame@48cad3fa3e0f1c669f859f2824312195',
  pw:channel:event   method: 'navigated',
  pw:channel:event   params: {
    pw:channel:event     url: 'https://app.<redacted>.com/new/admin//custom_offers/new/',
    pw:channel:event     name: '',
    pw:channel:event     error: 'net::ERR_ABORTED; maybe frame was detached?',
    pw:channel:event     newDocument: { request: [Object] }
                                        pw:channel:event   }
    pw:channel:event } +32ms
    pw:channel:event {
  pw:channel:event   guid: 'browser-context@fed02734621c7d5e763a8de68dbb009c',
  pw:channel:event   method: 'requestFailed',
  pw:channel:event   params: {
    pw:channel:event     request: { guid: 'request@246d6fbe69e502d0a1acad07ce9f71b4' },
    pw:channel:event     failureText: 'net::ERR_ABORTED',
    pw:channel:event     responseEndTiming: -1,
    pw:channel:event     page: { guid: 'page@57308fb78210dc9fe4cef469b1f3872e' }
                                 pw:channel:event   }
    pw:channel:event } +0ms
  pw:error Error: Target closed
    pw:error     at CRSession.send (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crConnection.js:196:34)
    pw:error     at CRSession._sendMayFail (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crConnection.js:211:17)
    pw:error     at FrameSession._stopScreencast (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crPage.js:1004:59)
    pw:error     at FrameSession._stopVideoRecording (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crPage.js:986:16)
    pw:error     at Page.<anonymous> (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crPage.js:965:57)
    pw:error     at Object.onceWrapper (node:events:509:28)
    pw:error     at Page.emit (node:events:402:35)
    pw:error     at Page._didClose (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/page.js:158:10)
    pw:error     at CRPage.didClose (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crPage.js:197:16)
    pw:error     at CRBrowser._onDetachedFromTarget (/Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crBrowser.js:229:14)
    pw:error     at CRSession.emit (node:events:390:28)
    pw:error     at /Users/<redacted>/root/poc/e2e-tests/node_modules/playwright-core/lib/server/chromium/crConnection.js:225:14 +0ms
```

**Comment by refactoreric** [comment]
Hi @ddeogharkar , looking at the source code in the GitHub repository, it does not look like you are executing the beforeAll anywhere. Because it's called in a hooks.ts, but I don't see that file being included anywhere.

Even if you do import it, the page that is created there will not be available to the tests in the TopNews.test.ts, because both files are modules, so variables need to be explicitly exported and imported to access them across files.

If you want to log on before each test, the login logic from the hooks file can best be moved to a fixture, for example called LoggedOnPage. Then in the definition of the other page object fixtures you can specify the LoggedOnPage as the dependency, instead of the plain regular page. Alternatively you could override the 'page' fixture if you want absolutely all your tests to be pre-logged in. But probably there are cases where you want to specifically test unauthenticated state ;-)

Have a look at https://playwright.dev/docs/test-fixtures for some examples which show how a fixture can do things (setup some application state) before your test.

**Comment by ddeogharkar** [comment]
Hi @refactoreric, Thanks for the answer. I will try to implement that, Actually my motive is to login only once before all the test cases and my entire suite will run with that login, So I wanted to move login function in test.BeforeAll and remove test.beforeEach but as I have created a separate file for beforeAll its not linked to any test cases. Can you please guide me How to achieve this?
Thanks,
@ddeogharkar

**Comment by refactoreric** [comment]
Hi @ddeogharkar , maybe this approach is what you are looking for:
https://playwright.dev/docs/test-auth#reuse-signed-in-state

There are several approaches described on that page to avoid having to log in for each test.

**Comment by ddeogharkar** [comment]
Yes, this looks good solution. Lets see if it works as there are multiple auths involved in the application.

**Comment by ddeogharkar** [comment]
Hi @refactoreric I am getting below error if I use global.config.ts file 
=========================== logs ===========================
waiting for selector "input[name="user"]"
============================================================
    at globalSetup (C:\Automation\Protractor-Automation\Test-sampleProt\global-setup.ts:7:14)
    at Runner._run (C:\Automation\Protractor-Automation\Test-sampleProt\node_modules\@playwright\test\lib\runner.js:246:58) {
  name: 'TimeoutError'
}
npm ERR! Test failed.  See above for more details.

I have pushed my changes to the same repo.

**Comment by refactoreric** [comment]
@ddeogharkar , looks like the locator for the username is incorrect. From what I can see, the GitHub login page uses a field identified as input[name="login"] for the user name.

**Comment by Evilweed** [comment]
I've added debug code to Playwright framework and this is the action that is crashing playwright.
![CleanShot 2021-12-21 at 20 20 15@2x](https://user-images.githubusercontent.com/5477821/146986007-dcaab273-0838-4376-af16-5c3a7406fffd.png)
![CleanShot 2021-12-21 at 20 21 23@2x](https://user-images.githubusercontent.com/5477821/146986073-392f7580-fee2-445f-9d58-1d30cc1aa7ec.png)

**Comment by yury-s** [comment]
@ddeogharkar as mentioned above there is no `input[name="user"]` on the github page, you need ` input[name="login"]`. Can you please reduce the example that you have a provide a minimal repro with actual URLs so that we can run it?

> So previous log was when we were clicking on input button, that triggers downloading of missing JS to show date picking modal. We have used page.eval to set the input value without clicking input. This worked but we are still blocked as trying to send the form after clicking submit also crashes Playwright - with following log output:

This issue seems different from the originally reported one and makes this thread hard to follow. I'd suggest you file a separate bug so that we can follow up. The exception inside the dispatcher suggests that the message may arrive after the context has been closed but we need more details. I looked at the log you attached above and it seems that it just hangs waiting for an input element for e2e-user-utfiE9Tegv@test.com

**Comment by ddeogharkar** [comment]
@refactoreric , I have created new repo - https://github.com/ddeogharkar/practiceglobalsetup please refer this & try running from your end currently test is passing but browser is getting opened & closed immediately, I am suspecting it is not including the global-setup.ts file.

**Comment by refactoreric** [comment]
Hi @ddeogharkar , since this is more a learning story than an issue in Playwright, maybe it's better to for the Playwright team to close this issue and for us to continue the conversation(s) on Slack: https://aka.ms/playwright-slack

I did have a quick look at your globalsetup repository. WARNING: Never save storageState.json in a Git repository shared with others.

If you want to check whether the globalSetup is called or not, you could try the [vscode JavaScript debugger](https://code.visualstudio.com/docs/nodejs/nodejs-debugging). 
Another way is [recording a trace file](https://playwright.dev/docs/api/class-tracing) for globalSetup.

**Comment by ddeogharkar** [comment]
@refactoreric Will work, you can close this issue.
